### ISSUE TITLE
How does convolution happen?

### ISSUE THREAD

**Issue by myuanqing** [intel/mkl-dnn#576]
Hi.
I am reading about the convolution operation in pytorch which includes mkl-dnn as third-party library. I've found that torch.nn.conv2d has something to do with jit_* in mkl-dnn, however, I don't know how. How does the jit_avx2 do MAC(mul and add) operations? That's quite confusing.

**Comment by vpirogov** [comment]
HI @myuanqing,

DNNL uses just-in-time (JIT) code generation in implementations of some functions, including convolutions. The library detects instruction set available at the current system in runtime and generates machine code specific for this instruction set. The code generators are specialized on convolution input and output sizes, shape of weights, batch size, and data types. You can find implementations of the generators in [src/cpu/jit_*](https://github.com/intel/mkl-dnn/tree/master/src/cpu) files.

For Intel AVX2 instruction set fp32 convolutions use FMA to perform MACs. Here's [relevant piece of code](https://github.com/intel/mkl-dnn/blob/master/src/cpu/jit_avx2_conv_kernel_f32.cpp#L80) for fp32 forward convolution with non-trivial filter.

Is there something specific you are trying to accomplish or just looking to understand how DNNL works?

**Comment by myuanqing** [comment]
Hi  Vadim,

Thank you for your reply. To be honest, I am trying to modify the precision of computations in Pytorch.  I suppose it’s necessary for me to understand how the program works to know what to do for the modification.


________________________________
发件人: Vadim Pirogov <notifications@github.com>
发送时间: Wednesday, October 23, 2019 11:46:00 PM
收件人: intel/mkl-dnn <mkl-dnn@noreply.github.com>
抄送: Shaun the Sheep <myuanqing@outlook.com>; Mention <mention@noreply.github.com>
主题: Re: [intel/mkl-dnn] How does convolution happen? (#576)


HI @myuanqing<https://github.com/myuanqing>,

DNNL uses just-in-time (JIT) code generation in implementations of some functions, including convolutions. The library detects instruction set available at the current system in runtime and generates machine code specific for this instruction set. The code generators are specialized on convolution input and output sizes, shape of weights, batch size, and data types. You can find implementations of the generators in src/cpu/jit_*<https://github.com/intel/mkl-dnn/tree/master/src/cpu> files.

For Intel AVX2 instruction set fp32 convolutions use FMA to perform MACs. Here's relevant piece of code<https://github.com/intel/mkl-dnn/blob/master/src/cpu/jit_avx2_conv_kernel_f32.cpp#L80> for fp32 forward convolution with non-trivial filter.

Is there something specific you are trying to accomplish or just looking to understand how DNNL works?

―

**Comment by vpirogov** [comment]
Can you elaborate on the precision part?Are you looking to implement convolutions in lower precision for performance or this is something else?

**Comment by myuanqing** [comment]
In fact, I’m designing an AI-chip that uses 27bits for convolution and I want to modify the precision of Pytorch so that I can find out whether the hardware code is correct by looking at the differences between the results of AI-chip and the modified Pytorch.
I trace the code’torch.nn.Conv2d’ and get here. Is it possible to modify the precision in MKL-DNN?


发件人: Vadim Pirogov<mailto:notifications@github.com>
发送时间: 2019年10月24日 0:07
收件人: intel/mkl-dnn<mailto:mkl-dnn@noreply.github.com>
抄送: Shaun the Sheep<mailto:myuanqing@outlook.com>; Mention<mailto:mention@noreply.github.com>
主题: Re: [intel/mkl-dnn] How does convolution happen? (#576)


Can you elaborate on the precision part?Are you looking to implement convolutions in lower precision for performance or this is something else?

―

**Comment by emfomenk** [comment]
Sounds cool!

Changing accuracy in Intel MKL-DNN definitely not an easy task, given that we have a lot of different implementations that we dispatch at run-time (so there is no single point that you can change and be sure that all convolutions will use the changed algorithm).
For this 27-bit HW is intermediate multiplication happens in true f32 or some modified data type that has 27-bit? In other words is it just the data type that uses 27 bits (to store data) and all computation still in f32 or the entire pipeline uses the novel data type?
In the former case it might be sufficient to continue using f32 data type in PyTorch and Intel MKL-DNN, but simply round the data before invoking the convolution:
```
  dst_f32 <- conv(src_f32, wei_f32)

  ==>

  src_f32 <- use_only_27bit(src_f32)
  wei_f32 <- use_only_27bit(wei_f32)
  dst_f32 <- conv(src_f32, wei_f32)
  dst_f32 <- use_only_27bit(dst_f32)
```
However, in the latter case you indeed have to change the inner-most computations (`fmadd231ps`-like instructions that @vpirogov mentioned above). This might be quite complicated task; and for such kind experiments I would definitely suggest to start with more reference code (i.e. that is written mostly for convenience, ease of understanding and changing, rather than for the best possible performance). Once you confirm the accuracy for some basic topologies and want to have a confirmation for the broader set of cases, in which case you need to have faster emulation, you may invest more effort in modifying the optimized code like Intel MKL-DNN.

**Comment by myuanqing** [comment]
Oh, thanks a lot! I'm not using some novel data type but Uint8 for activation. Multification is implemented as shifting operation (6 bits at most).  Plus, the largest kernel size supported is 3 (H) * 3 (W) * 512 (C). So the precision is 27 bit ( 8 + 6 + 4 + 9 ).  In this way, I think it's a neccesity to change the inner-most computation.

**Comment by vpirogov** [comment]
Considering that you are using int8 activations the [int8 convolution implementation](https://github.com/intel/mkl-dnn/blob/master/src/cpu/jit_avx512_core_x8s8s32x_conv_kernel.cpp#L399) may be better starting point.

**Comment by myuanqing** [comment]
Thank you.

**Comment by vpirogov** [comment]
Closing for now. Feel free to reopen or submit a new one if you have any further questions.
